File: py/src/imbrace/app/resources/ai.py
from typing import Any, AsyncIterator, Dict, Iterator, List, Optional
import json
from ...core.http import HttpTransport, AsyncHttpTransport


class AiResource:
    """AI domain — Sync. Completions + embeddings (+ streaming)."""

    def __init__(self, http: HttpTransport, base: str):
        self._http = http
        self._base = f"{base}/ai"

    def complete(self, model: str, messages: List[Dict[str, str]],
                 temperature: Optional[float] = None, max_tokens: Optional[int] = None,
                 metadata: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        body: Dict[str, Any] = {"model": model, "messages": messages, "stream": False}
        if temperature is not None:
            body["temperature"] = temperature
        if max_tokens is not None:
            body["max_tokens"] = max_tokens
        if metadata:
            body["metadata"] = metadata
        return self._http.request("POST", f"{self._base}/completions", json=body).json()

    def stream(self, model: str, messages: List[Dict[str, str]],
               temperature: Optional[float] = None, max_tokens: Optional[int] = None) -> Iterator[Dict[str, Any]]:
        """Streaming completion — yields SSE chunks."""
        import httpx
        body: Dict[str, Any] = {"model": model, "messages": messages, "stream": True}
        if temperature is not None:
            body["temperature"] = temperature
        if max_tokens is not None:
            body["max_tokens"] = max_tokens

        headers: Dict[str, str] = {"Content-Type": "application/json", "Accept": "text/event-stream"}
        if self._http.api_key:
            headers[self._http.token_header] = self._http.api_key
        token = self._http.token_manager.get_token()
        if token:
            headers[self._http.token_header] = token

        with httpx.Client(timeout=None) as client:
            with client.stream("POST", f"{self._base}/completions", json=body, headers=headers) as res:
                for line in res.iter_lines():
                    if line.startswith("data: "):
                        data = line[6:]
                        if data == "[DONE]":
                            break
                        try:
                            yield json.loads(data)
                        except json.JSONDecodeError:
                            continue

    def embed(self, model: str, input: List[str]) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._base}/embeddings",
                                  json={"model": model, "input": input}).json()


class AsyncAiResource:
    """AI domain — Async."""

    def __init__(self, http: AsyncHttpTransport, base: str):
        self._http = http
        self._base = f"{base}/ai"

    async def complete(self, model: str, messages: List[Dict[str, str]],
                       temperature: Optional[float] = None, max_tokens: Optional[int] = None,
                       metadata: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        body: Dict[str, Any] = {"model": model, "messages": messages, "stream": False}
        if temperature is not None:
            body["temperature"] = temperature
        if max_tokens is not None:
            body["max_tokens"] = max_tokens
        if metadata:
            body["metadata"] = metadata
        res = await self._http.request("POST", f"{self._base}/completions", json=body)
        return res.json()

    async def stream(self, model: str, messages: List[Dict[str, str]],
                     temperature: Optional[float] = None,
                     max_tokens: Optional[int] = None) -> AsyncIterator[Dict[str, Any]]:
        import httpx
        body: Dict[str, Any] = {"model": model, "messages": messages, "stream": True}
        if temperature is not None:
            body["temperature"] = temperature
        if max_tokens is not None:
            body["max_tokens"] = max_tokens

        headers: Dict[str, str] = {"Content-Type": "application/json", "Accept": "text/event-stream"}
        if self._http.api_key:
            headers[self._http.token_header] = self._http.api_key
        token = self._http.token_manager.get_token()
        if token:
            headers[self._http.token_header] = token

        async with httpx.AsyncClient(timeout=None) as client:
            async with client.stream("POST", f"{self._base}/completions", json=body, headers=headers) as res:
                async for line in res.aiter_lines():
                    if line.startswith("data: "):
                        data = line[6:]
                        if data == "[DONE]":
                            break
                        try:
                            yield json.loads(data)
                        except json.JSONDecodeError:
                            continue

    async def embed(self, model: str, input: List[str]) -> Dict[str, Any]:
        res = await self._http.request("POST", f"{self._base}/embeddings",
                                       json={"model": model, "input": input})
        return res.json()

File: ts/src/app/resources/ai.ts
import { HttpTransport } from "../../core/http.js"
import type { Completion, Embedding, CompletionInput, EmbeddingInput, StreamChunk } from "../../types/index.js"

export class AiResource {
  constructor(private readonly http: HttpTransport, private readonly base: string) {}

  private get url() { return `${this.base}/ai` }

  /**
   * Full (non-streaming) completion — waits for entire response.
   */
  async complete(input: CompletionInput): Promise<Completion> {
    return this.http.getFetch()(`${this.url}/completions`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ ...input, stream: false }),
    }).then(r => r.json())
  }

  /**
   * Streaming completion via SSE — returns an AsyncGenerator of StreamChunk.
   *
   * @example
   * for await (const chunk of client.ai.stream({ model: "gpt-4o", messages: [...] })) {
   *   process.stdout.write(chunk.choices[0]?.delta?.content ?? "")
   * }
   */
  async *stream(input: Omit<CompletionInput, "stream">): AsyncGenerator<StreamChunk> {
    const res = await this.http.getFetch()(`${this.url}/completions`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "Accept": "text/event-stream",
      },
      body: JSON.stringify({ ...input, stream: true }),
    })

    if (!res.body) throw new Error("No response body for streaming")

    const reader = res.body.getReader()
    const decoder = new TextDecoder()
    let buffer = ""

    try {
      while (true) {
        const { done, value } = await reader.read()
        if (done) break

        buffer += decoder.decode(value, { stream: true })
        const lines = buffer.split("\n")
        buffer = lines.pop() ?? ""

        for (const line of lines) {
          if (!line.startsWith("data: ")) continue
          const data = line.slice(6).trim()
          if (data === "[DONE]") return
          try {
            yield JSON.parse(data) as StreamChunk
          } catch {
            // skip malformed JSON lines
          }
        }
      }
    } finally {
      reader.releaseLock()
    }
  }

  /**
   * Generate embeddings for a list of strings.
   */
  async embed(input: EmbeddingInput): Promise<Embedding> {
    return this.http.getFetch()(`${this.url}/embeddings`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(input),
    }).then(r => r.json())
  }
}
